File: library/src/main/java/com/harvey/arcface/DetectFaceAction.java
package com.harvey.arcface;

import androidx.annotation.MainThread;
import android.util.Log;

import com.harvey.arcface.moodel.FaceFindModel;
import com.harvey.arcface.utils.MainHandler;

import java.util.List;

/**
 * Created by harvey on 2018/2/7 0007 09:45
 */

public class DetectFaceAction implements Runnable {
    final String TAG = "DetectFace";
    byte[] frameBytes;
    volatile boolean setToStop = false;
    volatile boolean isDetectingFace = false;
    int cameraWidth;
    int cameraHeight;
    OnFaceDetectListener faceDetectListener;

    public DetectFaceAction() {
    }

    public void destroy() {
        setToStop = true;
        frameBytes = null;
        faceDetectListener = null;
        isDetectingFace = false;
    }

    public void setData(byte[] bytes, int width, int height) {
        this.frameBytes = bytes;
        this.cameraWidth = width;
        this.cameraHeight = height;
    }

    public void setOnFaceDetectListener(OnFaceDetectListener l) {
        this.faceDetectListener = l;
    }

    @Override
    public void run() {
        while (!setToStop) {
            if (frameBytes != null && !isDetectingFace) {
                isDetectingFace = true;
                final List<FaceFindModel> faceFindModels = FaceManager.getInstance().extractAllFaceFeature(frameBytes, cameraWidth,
                        cameraHeight);
                callOnFaceDetect(faceFindModels);
                isDetectingFace = false;
            }
        }
    }

    void callOnFaceDetect(final List<FaceFindModel> data) {
        if (data.size() > 0)
            Log.d(TAG, "检测人脸数量-->" + data.size());
        MainHandler.run(new Runnable() {
            @Override
            public void run() {
                if (faceDetectListener != null) {
                    faceDetectListener.onFaceDetect(data, frameBytes);
                }
            }
        });

    }

    public interface OnFaceDetectListener {
        @MainThread
        void onFaceDetect(List<FaceFindModel> faceFindModels, byte[] frameBytes);
    }
}

File: kotlin-library/src/main/java/com/harvey/arcface/DetectFaceAction.kt
package com.harvey.arcface

import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import android.support.annotation.MainThread
import android.util.Log
import com.harvey.arcface.model.FaceFindModel

/**
 * Created by hanhui on 2018/6/1 0001 10:49
 */
class DetectFaceAction : Runnable {

    private val TAG = "DetectFace"
    private var mDetectFaceHandler: Handler
    private var mDetectFaceLooper: Looper
    private var frameBytes: ByteArray? = null
    @Volatile
    private var isDetectingFace = false// 正在检测人脸
    @Volatile
    private var isInit = false
    private var cameraWidth: Int = 0
    private var cameraHeight: Int = 0
    private var faceDetectListener: OnFaceDetectListener? = null

    init {
        val detectFaceThread = HandlerThread("DetectFace")
        detectFaceThread.start()
        mDetectFaceLooper = detectFaceThread.looper
        mDetectFaceHandler = Handler(mDetectFaceLooper)
        isInit = true
    }

    fun destroy() {
        if (isInit) {
            mDetectFaceLooper.quit()
        }
        frameBytes = null
        isDetectingFace = false
        faceDetectListener = null
        isInit = false
    }

    fun detectFace(bytes: ByteArray, width: Int, height: Int) {
        if (!isInit)
            throw NullPointerException("没有初始化人脸检测!")
        if (isDetectingFace)
            return
        this.frameBytes = bytes
        this.cameraWidth = width
        this.cameraHeight = height
        mDetectFaceHandler.post(this)
    }

    fun setOnFaceDetectListener(l: OnFaceDetectListener) {
        this.faceDetectListener = l
    }

    override fun run() {
        if (isDetectingFace)
            return
        if (frameBytes == null)
            return
        isDetectingFace = true
        val faceFindModels = ArcFaceEngine.detectFace(frameBytes!!, cameraWidth,
                cameraHeight)
        callOnFaceDetect(faceFindModels)
        isDetectingFace = false
    }

    private fun callOnFaceDetect(data: List<FaceFindModel>) {
        if (data.isNotEmpty())
            Log.d(TAG, "检测人脸数量-->" + data.size)
        MainHandler.run(Runnable { faceDetectListener?.onFaceDetect(data, frameBytes!!) })
    }

    interface OnFaceDetectListener {
        @MainThread
        fun onFaceDetect(faceFindModels: List<FaceFindModel>, frameBytes: ByteArray)
    }
}